File: VulkanGameEngineLevelEditor/GameEngineAPI/Sprite.cs
using GlmSharp;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using VulkanGameEngineLevelEditor.Components;
using VulkanGameEngineLevelEditor.Models;

namespace VulkanGameEngineLevelEditor.GameEngineAPI
{
    public class Sprite
    {
        [JsonIgnore]
        public GameObject ParentGameObject { get; set; }

        [JsonIgnore]
        public JsonPipeline<Vertex2D> renderPipeline { get; set; }

        [JsonIgnore]
        public Transform2DComponent Transform2D { get; set; }

        [JsonIgnore]
        public uint CurrentAnimationID { get; set; } = 0;

        [JsonIgnore]
        public uint CurrentFrame { get; set; } = 0;

        [JsonIgnore]
        public bool SpriteAlive { get; set; } = true;

        [JsonIgnore]
        public Animation2D CurrentSpriteAnimation { get; set; }

        [JsonIgnore]
        public Material SpriteMaterial { get; set; }

        [JsonIgnore]
        public SpriteInstanceStruct SpriteInstance;

        public vec2 SpriteSize { get; set; } = new vec2(50.0f);
        public vec4 SpriteColor { get; set; } = new vec4(0.0f, 0.0f, 0.0f, 1.0f);
        public vec2 SpritePosition { get; set; } = new vec2(0.0f);
        public vec2 SpriteRotation { get; set; } = new vec2(0.0f);
        public vec2 SpriteScale { get; set; } = new vec2(1.0f);
        public uint SpriteLayer { get; set; } = 0;
        public SpriteSheet Spritesheet { get; set; }
        public List<Animation2D> AnimationList { get; set; } = new List<Animation2D>();
        public Guid SpriteMaterialID
        {
            get
            {
                return SpriteMaterial.MaterialID;
            }
        }

        Sprite()
        {

        }

        public Sprite(GameObject parentGameObject, SpriteModel model)
        {
            ParentGameObject = parentGameObject;
            Transform2D = (Transform2DComponent)ParentGameObject.GetComponentByComponentType(ComponentTypeEnum.kTransform2DComponent);

            Spritesheet = model.Spritesheet;
            SpriteSize = model.SpriteSize;
            SpriteColor = model.SpriteColor;
            SpritePosition = model.SpritePosition;
            SpriteRotation = model.SpriteRotation;
            SpriteScale = model.SpriteScale;
            SpriteLayer = model.SpriteLayer;
            SpriteMaterial = renderPipeline._GPUImport.MaterialList.Where(x => x.MaterialID == SpriteMaterialID).First();

            AnimationList = new List<Animation2D>();
            foreach (var animation in model.AnimationList)
            {
                AnimationList.Add(new Animation2D(animation));
            }
            CurrentSpriteAnimation = AnimationList.Last();
        }

        public Sprite(GameObject parentObject, SpriteSheet spriteSheet)
        {
            ParentGameObject = parentObject;
            Transform2D = (Transform2DComponent)ParentGameObject.GetComponentByComponentType(ComponentTypeEnum.kTransform2DComponent);

            Spritesheet = spriteSheet;
            SpriteLayer = spriteSheet.SpriteLayer;
            SpriteSize = new vec2(spriteSheet.SpritePixelSize.x * spriteSheet.SpriteScale.x, spriteSheet.SpritePixelSize.y * spriteSheet.SpriteScale.y);
            SpriteColor = new vec4(0.0f, 0.0f, 0.0f, 1.0f);
            SpriteMaterial = spriteSheet.SpriteMaterial;
            SpriteInstance = new SpriteInstanceStruct();

            List<ivec2> frameList = new List<ivec2>
            {
                new ivec2(0, 0),
                new ivec2(1, 0)
            };

            AnimationList.Add(new Animation2D("Standing", frameList, 0.2f));

            List<ivec2> frameList2 = new List<ivec2>
            {
               new ivec2(3, 0),
                new ivec2(4, 0),
                new ivec2(5, 0),
                new ivec2(4, 0)
            };
            AnimationList.Add(new Animation2D("Walking", frameList2, 0.5f));
            CurrentSpriteAnimation = AnimationList.Last();
        }

        public void Input(float deltaTime)
        {

        }

        public void Update(VkCommandBuffer commandBuffer, float deltaTime)
        {
            if (Transform2D != null)
            {
                mat4 spriteMatrix = mat4.Identity;
                spriteMatrix = mat4.Translate(new vec3(Transform2D.GameObjectPosition.x, Transform2D.GameObjectPosition.y, 0.0f));
                spriteMatrix = mat4.Rotate(VMath.DegreesToRadians(Transform2D.GameObjectRotation.x), new vec3(1.0f, 0.0f, 0.0f));
                spriteMatrix = mat4.Rotate(VMath.DegreesToRadians(Transform2D.GameObjectRotation.y), new vec3(0.0f, 1.0f, 0.0f));
                spriteMatrix = mat4.Rotate(VMath.DegreesToRadians(0.0f), new vec3(0.0f, 0.0f, 1.0f));
                spriteMatrix = mat4.Scale(new vec3(Transform2D.GameObjectScale.x, Transform2D.GameObjectScale.y, 0.0f));


                SpriteInstance.SpritePosition = Transform2D.GameObjectPosition;
                SpriteInstance.SpriteSize = SpriteSize;
                SpriteInstance.UVOffset = new vec4(Spritesheet.SpriteUVSize.x * CurrentSpriteAnimation.FrameList[(int)CurrentFrame].x, Spritesheet.SpriteUVSize.y * CurrentSpriteAnimation.FrameList[(int)CurrentFrame].y, Spritesheet.SpriteUVSize.x, Spritesheet.SpriteUVSize.y);
                SpriteInstance.Color = SpriteColor;
                SpriteInstance.MaterialID = (SpriteMaterial != null) ? SpriteMaterial.MaterialBufferIndex : 0;
                SpriteInstance.InstanceTransform = spriteMatrix;
            }

            CurrentSpriteAnimation.CurrentFrameTime += deltaTime;
            if (CurrentSpriteAnimation.CurrentFrameTime >= CurrentSpriteAnimation.FrameHoldTime)
            {
                CurrentFrame += 1;
                CurrentSpriteAnimation.CurrentFrameTime = 0.0f;
                if (CurrentFrame > CurrentSpriteAnimation.FrameList.Count() - 1)
                {
                    CurrentFrame = 0;
                }
            }
        }

        public void Destroy()
        {

        }
    }
}

File: VulkanEngineDLL/Sprite.cpp
#include "Sprite.h"

void Sprite_BatchSpriteUpdate(SpriteInstanceStruct* spriteInstanceList, Sprite* spriteList, const Transform2DComponent* transform2DList, const SpriteVram* vramList, const Animation2D* animationList, const AnimationFrames* frameList, const Material* materialList, size_t spriteCount, float deltaTime)
{
    for (size_t x = 0; x < spriteCount; x++)
    {
        glm::mat4 spriteMatrix = glm::mat4(1.0f);
        if (spriteList[x].LastSpritePosition != spriteList[x].SpritePosition) {
            spriteMatrix = glm::translate(spriteMatrix, glm::vec3(transform2DList[x].GameObjectPosition.x, transform2DList[x].GameObjectPosition.y, 0.0f));
            spriteList[x].LastSpritePosition = spriteList[x].SpritePosition;
        }
        if (spriteList[x].LastSpriteRotation != spriteList[x].SpriteRotation) {
            spriteMatrix = glm::rotate(spriteMatrix, glm::radians(transform2DList[x].GameObjectRotation.x), glm::vec3(1.0f, 0.0f, 0.0f));
            spriteMatrix = glm::rotate(spriteMatrix, glm::radians(transform2DList[x].GameObjectRotation.y), glm::vec3(0.0f, 1.0f, 0.0f));
            spriteMatrix = glm::rotate(spriteMatrix, glm::radians(0.0f), glm::vec3(0.0f, 0.0f, 1.0f));
            spriteList[x].LastSpriteRotation = spriteList[x].SpriteRotation;
        }
        if (spriteList[x].LastSpriteScale != spriteList[x].SpriteScale) {
            spriteMatrix = glm::scale(spriteMatrix, glm::vec3(transform2DList[x].GameObjectScale.x, transform2DList[x].GameObjectScale.y, 1.0f));
            spriteList[x].LastSpriteScale = spriteList[x].SpriteScale;
        }

        spriteList[x].CurrentFrameTime += deltaTime;
        if (spriteList[x].CurrentFrameTime >= animationList[x].FrameHoldTime)
        {
            spriteList[x].CurrentFrame += 1;
            spriteList[x].CurrentFrameTime = 0.0f;
            if (spriteList[x].CurrentFrame >= spriteCount)
            {
                spriteList[x].CurrentFrame = 0;
            }
        }

        const ivec2& currentFrame = frameList[x][spriteList[x].CurrentFrame];
        spriteInstanceList[x].SpritePosition = transform2DList[x].GameObjectPosition;
        spriteInstanceList[x].SpriteSize = vramList[x].SpriteSize;
        spriteInstanceList[x].MaterialID = materialList[x].ShaderMaterialBufferIndex;
        spriteInstanceList[x].InstanceTransform = spriteMatrix;
        spriteInstanceList[x].FlipSprite = spriteList[x].FlipSprite;
        spriteInstanceList[x].UVOffset = glm::vec4(vramList[x].SpriteUVSize.x * currentFrame.x, vramList[x].SpriteUVSize.y * currentFrame.y, vramList[x].SpriteUVSize.x, vramList[x].SpriteUVSize.y);
    }
}

SpriteInstanceStruct Sprite_PerSpriteUpdate(const Transform2DComponent& transform2D, const SpriteVram& vram, const Animation2D& animation, const AnimationFrames& frameList, const Material& material, const ivec2& currentFrame, Sprite& sprite, float deltaTime)
{
    mat4 spriteMatrix = mat4(1.0f);
    if (sprite.LastSpritePosition != sprite.SpritePosition)
    {
        spriteMatrix = glm::translate(spriteMatrix, vec3(transform2D.GameObjectPosition.x, transform2D.GameObjectPosition.y, 0.0f));
        sprite.LastSpritePosition == sprite.SpritePosition;
    }
    if (sprite.LastSpriteRotation != sprite.SpriteRotation)
    {
        spriteMatrix = glm::rotate(spriteMatrix, glm::radians(transform2D.GameObjectRotation.x), vec3(1.0f, 0.0f, 0.0f));
        spriteMatrix = glm::rotate(spriteMatrix, glm::radians(transform2D.GameObjectRotation.y), vec3(0.0f, 1.0f, 0.0f));
        spriteMatrix = glm::rotate(spriteMatrix, glm::radians(0.0f), vec3(0.0f, 0.0f, 1.0f));
        sprite.LastSpriteRotation == sprite.SpriteRotation;
    }
    if (sprite.LastSpriteScale != sprite.SpriteScale)
    {
        spriteMatrix = glm::scale(spriteMatrix, vec3(transform2D.GameObjectScale.x, transform2D.GameObjectScale.y, 1.0f));
        sprite.LastSpriteScale == sprite.SpriteScale;
    }

    sprite.CurrentFrameTime += deltaTime;
    if (sprite.CurrentFrameTime >= animation.FrameHoldTime) {
        sprite.CurrentFrame += 1;
        sprite.CurrentFrameTime = 0.0f;
        if (sprite.CurrentFrame >= frameList.size())
        {
            sprite.CurrentFrame = 0;
        }
    }

    SpriteInstanceStruct spriteInstance;
    spriteInstance.SpritePosition = transform2D.GameObjectPosition;
    spriteInstance.SpriteSize = vram.SpriteSize;
    spriteInstance.MaterialID = material.ShaderMaterialBufferIndex;
    spriteInstance.InstanceTransform = spriteMatrix;
    spriteInstance.FlipSprite = sprite.FlipSprite;
    spriteInstance.UVOffset = vec4(vram.SpriteUVSize.x * currentFrame.x, vram.SpriteUVSize.y * currentFrame.y, vram.SpriteUVSize.x, vram.SpriteUVSize.y);

    return spriteInstance;
}

void Sprite_SetSpriteAnimation(Sprite& sprite, Sprite::SpriteAnimationEnum spriteAnimation)
{
    if (sprite.CurrentAnimationID == spriteAnimation)
    {
        return;
    }

    sprite.CurrentAnimationID = spriteAnimation;
    sprite.CurrentFrame = 0;
    sprite.CurrentFrameTime = 0.0f;
}
